worlds/zillion/options.py

```diff
@@ -0,0 +1,377 @@
+
+    Vanilla is 4
+
+    maximum total for all items is 144
+    """
+    range_start = 0
+    range_end = 126
+    default = 4
+    display_name = "Scope count"
+
+
+class ZillionRedIDCardCount(Range):
+    """
+    how many Red ID Cards are in the game
+
+    Vanilla is 1
+
+    maximum total for all items is 144
+    """
+    range_start = 0
+    range_end = 126
+    default = 2
+    display_name = "Red ID Card count"
+
+
+class ZillionSkill(Range):
+    """ the difficulty level of the game """
+    range_start = 0
+    range_end = 5
+    default = 2
+
+
+class ZillionStartingCards(SpecialRange):
+    """
+    how many ID Cards to start the game with
+
+    Refilling at the ship also ensures you have at least this many cards.
+    0 gives vanilla behavior.
+    """
+    default = 2
+    range_start = 0
+    range_end = 10
+    display_name = "starting cards"
+    special_range_names = {
+        "vanilla": 0
+    }
+
+
+class ZillionRoomGen(Toggle):
+    """ whether to generate rooms with random terrain """
+    display_name = "room generation"
+
+
+zillion_options: Dict[str, AssembleOptions] = {
+    "continues": ZillionContinues,
+    # "early_scope": ZillionEarlyScope,  # TODO: implement
+    "floppy_req": ZillionFloppyReq,
+    "gun_levels": ZillionGunLevels,
+    "jump_levels": ZillionJumpLevels,
+    "randomize_alarms": ZillionRandomizeAlarms,
+    "max_level": ZillionMaxLevel,
+    "start_char": ZillionStartChar,
+    "opas_per_level": ZillionOpasPerLevel,
+    "id_card_count": ZillionIDCardCount,
+    "bread_count": ZillionBreadCount,
+    "opa_opa_count": ZillionOpaOpaCount,
+    "zillion_count": ZillionZillionCount,
+    "floppy_disk_count": ZillionFloppyDiskCount,
+    "scope_count": ZillionScopeCount,
+    "red_id_card_count": ZillionRedIDCardCount,
+    "skill": ZillionSkill,
+    "starting_cards": ZillionStartingCards,
+    "room_gen": ZillionRoomGen,
+}
+
+
+def convert_item_counts(ic: "Counter[str]") -> ZzItemCounts:
+    tr: ZzItemCounts = {
+        ID.card: ic["ID Card"],
+        ID.red: ic["Red ID Card"],
+        ID.floppy: ic["Floppy Disk"],
+        ID.bread: ic["Bread"],
+        ID.gun: ic["Zillion"],
+        ID.opa: ic["Opa-Opa"],
+        ID.scope: ic["Scope"],
+        ID.empty: ic["Empty"],
+    }
+    return tr
+
+
+def validate(world: "MultiWorld", p: int) -> "Tuple[ZzOptions, Counter[str]]":
+    """
+    adjusts options to make game completion possible
+
+    `world` parameter is MultiWorld object that has my options on it
+    `p` is my player id
+    """
+    for option_name in zillion_options:
+        assert hasattr(world, option_name), f"Zillion option {option_name} didn't get put in world object"
+    wo = cast(Any, world)  # so I don't need getattr on all the options
+
+    jump_levels = cast(ZillionJumpLevels, wo.jump_levels[p])
+    jump_option = jump_levels.get_current_option_name().lower()
+    assert jump_option in VBLR_CHOICES, f"{jump_option} in {VBLR_CHOICES}"
```

Comment:
while it does not hurt, options only having valid values is something that should be guaranteed by the options system. I think having a test for `all VBLR options in VBLR_CHOICES` and removing this assert would be a better split. (but tests are still TODO, right?)

Reply:
This assert isn't checking for a valid value of the Archipelago option. It's checking for a valid value for the zilliandomizer option, just in case something changes on one and not the other.
But yeah, it could be done in unit tests.

Is there an established pattern in the tests for iterating through all the possible values of an option.

Reply:
`for name in Option.name_lookup.values():`

Reply:
now that it's in tests - do we want to get rid of the assert? It's not like an assert that's run once per generation is harmful to performance, but too many checks and asserts makes it hard to read code imo. Should the options system be flawed, (multiple games) will hopefully crash at a later stage of generation anyway.

Reply:
I don't yet have a test to process my zilliandomizer options in the zilliandomizer system from all of the Archipelago option values.

The test that I have doesn't feel right because it depends on the implementation in the test being the same as the implementation in runtime code.

Both have `.lower()` on the string. But if the test has `.lower()` and the runtime code gets changed to not have `.lower()`, then the test passes while the runtime code fails. (And this is a detail that would not be caught by other games.)